frontend/model/notifications/mainPeriodicNotificationEntries.js

```diff
@@ -0,0 +1,240 @@
+                  proposalData: proposal.data.proposalData
+                })
+              }
+            }
+          }
+
+          return { contractID, expiringProposals, groupNotificationItems, expiredProposalIds }
+        }).filter(entry => entry.expiringProposals.length || entry.groupNotificationItems.length || entry.expiredProposalIds.length)
+
+        return this.expiringOrExpiredProposalsByGroup.length
+      },
+      async emit () {
+        for (const { contractID, expiringProposals, groupNotificationItems, expiredProposalIds } of this.expiringOrExpiredProposalsByGroup) {
+          if (expiringProposals.length) {
+            await sbp('gi.actions/group/notifyExpiringProposals', {
+              contractID,
+              data: { proposals: expiringProposals }
+            })
+          }
+
+          if (groupNotificationItems.length) {
+            groupNotificationItems.forEach(proposal => {
+              sbp('gi.notifications/emit', 'PROPOSAL_EXPIRING', {
+                groupID: contractID,
+                creatorID: proposal.creatorID,
+                proposalId: proposal.proposalId,
+                proposalType: proposal.proposalType,
+                proposalData: proposal.proposalData,
+                title: proposal.proposalType === PROPOSAL_GENERIC ? proposal.proposalData.name : ''
+              })
+            })
+          }
+
+          if (expiredProposalIds.length) {
+            sbp('gi.actions/group/markProposalsExpired', {
+              contractID,
+              data: { proposalIds: expiredProposalIds }
+            })
+              .catch((e) => {
+                console.error('Error calling markProposalsExpired from notifications mixin', e)
+              })
+          }
+        }
+      },
+      shouldClearStateKey: () => true
+    }
+  },
+  {
+    type: PERIODIC_NOTIFICATION_TYPE.MIN5,
+    notificationData: {
+      stateKey: 'lastLoggedIn',
+      emitCondition ({ rootGetters }) {
+        return !!rootGetters.ourIdentityContractId
+      },
+      emit ({ rootState, rootGetters }) {
+        Promise.all(
+          rootGetters.groupsByName.filter(({ active }) => active).map(({ contractID }) => {
+            return sbp('gi.actions/group/kv/updateLastLoggedIn', {
+              contractID,
+              throttle: true
+            })
+          })
+        ).catch((e) => {
+          console.error('Error updating lastLoggedIn', e)
+        })
+      },
+      shouldClearStateKey: () => true
+    }
+  },
+  {
+    type: PERIODIC_NOTIFICATION_TYPE.MIN1,
+    notificationData: {
+      stateKey: 'reattemptFailedJoins',
+      emitCondition ({ rootGetters }) {
+        return !!rootGetters.ourIdentityContractId
+      },
+      emit () {
+        sbp('gi.actions/group/reattemptFailedJoins').catch((e) => {
+          console.error('Error calling reattemptFailedJoins', e)
+        })
+      },
+      shouldClearStateKey: () => true
+    }
+  }
```

Comment:
I see that the logic for `'gi.actions/group/reattemptFailedJoins'` works, but I'm not thrilled with the idea of having the app always run it every single minute.

Is there a way we can either do this differently, or modify the `emitCondition` to return true only when there is at least one group that we failed to join?

Reply:
I don't see the change request, but:

> Is there a way we can either do this differently

There almost always is, but differently how? I can't see how to possibly to this 're-trying' without re-attempting regularly. The one minute threshold was chosen because it's relatively quick. If you got stuck at the group join screen for 15 minutes, you'd (rightfully) think that something's broken.

> modify the emitCondition to return true only when there is at least one group that we failed to join?

That'd still run every minute, and `reattemptFailedJoins` already is a no-op if there's nothing to retry.

Reply:
> There almost always is, but differently how? I can't see how to possibly to this 're-trying' without re-attempting regularly. 

Well, for example, we could add a feature to add and remove periodic notifications based on whether or not we know they're necessary, or do something that has the safe effect.

There's also the concern that with this done as a periodic notification, `'gi.actions/group/reattemptFailedJoins'` will get called multiple times, once with a run on a slow connection to join the group, and while the join process is waiting, it gets called yet again because of the slow connection... Would that result in problems?

Reply:
If you'd really like to, I could implement this entirely within `group.js`, by using a `setTimeout` in there, clearing the timeout when not needed and avoiding concurrent calls. I don't think the extra complexity is worth it, but it's an option.

Reply:
That might be a good idea, yeah. If it turns out to be possible to do it in a way that's not super complex, then yeah let's do the `setTimeout` method instead. Also, if the timeout somehow gets cleared (e.g. if the SW is unloaded or w/e and the timeouts wiped), then can you make sure that if the join link is visited again we still recover?